common/src/main/java/com/microsoft/identity/common/internal/eststelemetry/EstsTelemetry.java

```diff
@@ -321,17 +363,63 @@ String getCurrentTelemetryHeaderString() {
+        if (mLastRequestTelemetryCache == null) {
             return null;
         }
 
-        return lastTelemetry.getCompleteTelemetryHeaderString();
+        LastRequestTelemetry lastRequestTelemetry = (LastRequestTelemetry) mLastRequestTelemetryCache.getRequestTelemetryFromCache();
+
+        if (lastRequestTelemetry == null) {
+            // we did not have anything in the telemetry cache for the last request
+            // let's create a new object based on the data available from the current request object
+            // and return the header string formed via that object
+            final String correlationId = DiagnosticContext.getRequestContext().get(DiagnosticContext.CORRELATION_ID);
+            final CurrentRequestTelemetry currentRequestTelemetry = mTelemetryMap.get(correlationId);
+            lastRequestTelemetry = new LastRequestTelemetry(currentRequestTelemetry.getSchemaVersion());
+            lastRequestTelemetry = (LastRequestTelemetry) lastRequestTelemetry.derive(currentRequestTelemetry);
+            return lastRequestTelemetry.getCompleteHeaderString();
+        }
+
+        // create a copy of the object retrieved from cache
+        LastRequestTelemetry lastRequestTelemetryCopy = new LastRequestTelemetry(lastRequestTelemetry.getSchemaVersion());
+        lastRequestTelemetryCopy = (LastRequestTelemetry) lastRequestTelemetryCopy.derive(lastRequestTelemetry);
+
+        // failed request data from the object retrieved from cache
+        final List<FailedRequest> originalFailedRequests = lastRequestTelemetry.getFailedRequests();
+
+        // error data from the object retrieved from cache
+        final List<String> originalErrors = lastRequestTelemetry.getErrors();
+
+        // the array index that marks how many elements were included in the header sent in the request
+        int indexOfFailedRequestsAndErrors = 0;
+
+        for (int i = 0; i < originalFailedRequests.size(); i++) {
+            // there is a limit of 8KB for the payload sent in request headers
+            // we will be maxing out at 4KB to avoid HTTP 413 errors
+            // check if we have enough space in the String to store another failed request/error element
+            // if yes, then add it to the failed request array (for the copy)
+            if (lastRequestTelemetryCopy.getCompleteHeaderString().length() < 3800) {
+                lastRequestTelemetryCopy.appendFailedRequestWithError(
+                        originalFailedRequests.get(indexOfFailedRequestsAndErrors),
+                        originalErrors.get(indexOfFailedRequestsAndErrors++)
+                );
+            } else {
+                // if there is no room for more data, then break out of this loop
+                break;
+            }
+        }
+
+        // tell cache where we stopped adding data to header so we can pick this up in the next
+        // network request
+        ((SharedPreferencesLastRequestTelemetryCache) mLastRequestTelemetryCache)
```

Comment:
I suggest keeping a separate copy of this inside the last request object (as same as the arrays).

Every telemetry values should be modified in only one place - flush().

Reply:
This field has been removed now per your suggestion.